File: BikeRoutePlanner.cs
// BikeRoutePlanner.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace BikeRoutePlanner
{
    class Program
    {
        static void Main(string[] args)
        {
            var opts = new Options();
            for (int i = 0; i < args.Length; i++)
            {
                switch (args[i])
                {
                    case "--start": opts.Start = args[++i]; break;
                    case "--end": opts.End = args[++i]; break;
                    case "--waypoints": opts.Waypoints = args[++i]; break;
                    case "--profile": opts.Profile = args[++i]; break;
                    case "--export-gpx": opts.GpxFile = args[++i]; break;
                    case "--save-json": opts.JsonFile = args[++i]; break;
                    case "--color": opts.Color = true; break;
                }
            }
            var planner = new Planner(opts);
            planner.Run();
        }

        class Options
        {
            public string Start { get; set; }
            public string End { get; set; }
            public string Waypoints { get; set; }
            public string Profile { get; set; } = "touring";
            public string GpxFile { get; set; }
            public string JsonFile { get; set; }
            public bool Color { get; set; }
        }

        class Point
        {
            public double Lat { get; set; }
            public double Lng { get; set; }
            public string Name { get; set; }
        }

        class Route
        {
            public Point Start { get; set; }
            public Point End { get; set; }
            public List<Point> Waypoints { get; set; }
            public double TotalDistanceKm { get; set; }
            public int ElevationGainM { get; set; }
            public double TimeHours { get; set; }
            public string Profile { get; set; }
            [JsonIgnore]
            public List<Point> Points { get; set; }
        }

        class Planner
        {
            private Options opts;
            private bool color;

            public Planner(Options opts)
            {
                this.opts = opts;
                this.color = opts.Color || !Console.IsOutputRedirected;
            }

            private Point ParseLocation(string loc)
            {
                if (loc.Contains(","))
                {
                    var parts = loc.Split(',');
                    if (double.TryParse(parts[0].Trim(), out double lat) && double.TryParse(parts[1].Trim(), out double lng))
                        return new Point { Lat = lat, Lng = lng, Name = loc };
                }
                return new Point { Lat = 0, Lng = 0, Name = loc };
            }

            private Point Geocode(string name)
            {
                int hash = Math.Abs(name.GetHashCode()) % 360;
                return new Point
                {
                    Lat = 50 + (hash % 100) / 100.0,
                    Lng = 10 + (hash % 200) / 100.0,
                    Name = name
                };
            }

            private double Haversine(double lat1, double lon1, double lat2, double lon2)
            {
                const double R = 6371.0;
                double phi1 = lat1 * Math.PI / 180;
                double phi2 = lat2 * Math.PI / 180;
                double dphi = (lat2 - lat1) * Math.PI / 180;
                double dlambda = (lon2 - lon1) * Math.PI / 180;
                double a = Math.Sin(dphi / 2) * Math.Sin(dphi / 2) +
                           Math.Cos(phi1) * Math.Cos(phi2) * Math.Sin(dlambda / 2) * Math.Sin(dlambda / 2);
                return R * 2 * Math.Atan2(Math.Sqrt(a), Math.Sqrt(1 - a));
            }

            private double GetSpeed()
            {
                return opts.Profile switch
                {
                    "road" => 25,
                    "mtb" => 15,
                    _ => 20
                };
            }

            public Route CalculateRoute()
            {
                var start = ParseLocation(opts.Start);
                var end = ParseLocation(opts.End);
                var wayPts = new List<Point>();
                if (!string.IsNullOrEmpty(opts.Waypoints))
                {
                    foreach (var w in opts.Waypoints.Split(','))
                        wayPts.Add(ParseLocation(w.Trim()));
                }
                if (start.Lat == 0) start = Geocode(start.Name);
                if (end.Lat == 0) end = Geocode(end.Name);
                foreach (var w in wayPts)
                    if (w.Lat == 0) { var g = Geocode(w.Name); w.Lat = g.Lat; w.Lng = g.Lng; }
                var points = new List<Point> { start };
                points.AddRange(wayPts);
                points.Add(end);
                double totalDist = 0;
                for (int i = 0; i < points.Count - 1; i++)
                    totalDist += Haversine(points[i].Lat, points[i].Lng, points[i + 1].Lat, points[i + 1].Lng);
                double speed = GetSpeed();
                double timeHours = totalDist / speed;
                int elevGain = (int)(totalDist * 0.05 * 1000);
                return new Route
                {
                    Start = start,
                    End = end,
                    Waypoints = wayPts,
                    TotalDistanceKm = totalDist,
                    ElevationGainM = elevGain,
                    TimeHours = timeHours,
                    Profile = opts.Profile,
                    Points = points
                };
            }

            public void PrintRoute(Route r)
            {
                if (color)
                {
                    Console.ForegroundColor = ConsoleColor.Cyan;
                    Console.WriteLine($"🚴 Маршрут от {r.Start.Name} до {r.End.Name}");
                    if (r.Waypoints.Any())
                    {
                        Console.ForegroundColor = ConsoleColor.Magenta;
                        Console.WriteLine($"   через {string.Join(", ", r.Waypoints.Select(w => w.Name))}");
                    }
                    Console.ForegroundColor = ConsoleColor.Green;
                    Console.WriteLine($"Расстояние: {r.TotalDistanceKm:F1} км");
                    Console.ForegroundColor = ConsoleColor.Yellow;
                    Console.WriteLine($"Набор высоты: {r.ElevationGainM} м");
                    Console.ForegroundColor = ConsoleColor.Blue;
                    Console.WriteLine($"Время в пути: ~{r.TimeHours:F1} ч");
                    Console.ResetColor();
                }
                else
                {
                    Console.WriteLine($"Маршрут от {r.Start.Name} до {r.End.Name}");
                    if (r.Waypoints.Any())
                        Console.WriteLine($"   через {string.Join(", ", r.Waypoints.Select(w => w.Name))}");
                    Console.WriteLine($"Расстояние: {r.TotalDistanceKm:F1} км");
                    Console.WriteLine($"Набор высоты: {r.ElevationGainM} м");
                    Console.WriteLine($"Время в пути: ~{r.TimeHours:F1} ч");
                }
            }

            public void ExportGPX(Route r, string filename)
            {
                var gpx = new System.Text.StringBuilder();
                gpx.AppendLine("<?xml version=\"1.0\" encoding=\"UTF-8\"?>");
                gpx.AppendLine("<gpx version=\"1.1\" creator=\"BikeRoutePlanner\">");
                gpx.AppendLine("  <trk>");
                gpx.AppendLine("    <trkseg>");
                foreach (var pt in r.Points)
                {
                    gpx.AppendLine($"      <trkpt lat=\"{pt.Lat:F6}\" lon=\"{pt.Lng:F6}\">");
                    gpx.AppendLine($"        <ele>{(int)(50 + pt.Lat * 10)}</ele>");
                    gpx.AppendLine("      </trkpt>");
                }
                gpx.AppendLine("    </trkseg>");
                gpx.AppendLine("  </trk>");
                gpx.AppendLine("</gpx>");
                File.WriteAllText(filename, gpx.ToString());
                Console.WriteLine($"GPX экспортирован в {filename}");
            }

            public void SaveJSON(Route r, string filename)
            {
                var options = new JsonSerializerOptions { WriteIndented = true };
                var json = JsonSerializer.Serialize(r, options);
                File.WriteAllText(filename, json);
                Console.WriteLine($"Маршрут сохранён в {filename}");
            }

            public void Run()
            {
                var route = CalculateRoute();
                PrintRoute(route);
                if (!string.IsNullOrEmpty(opts.GpxFile)) ExportGPX(route, opts.GpxFile);
                if (!string.IsNullOrEmpty(opts.JsonFile)) SaveJSON(route, opts.JsonFile);
            }
        }
    }
}

File: BikeRoutePlanner.java
// BikeRoutePlanner.java
import com.beust.jcommander.JCommander;
import com.beust.jcommander.Parameter;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import okhttp3.*;

import java.io.*;
import java.nio.file.*;
import java.util.*;

public class BikeRoutePlanner {
    @Parameter(names = "--start", required = true)
    private String start;
    @Parameter(names = "--end", required = true)
    private String end;
    @Parameter(names = "--waypoints")
    private String waypoints;
    @Parameter(names = "--profile")
    private String profile = "touring";
    @Parameter(names = "--export-gpx")
    private String gpxFile;
    @Parameter(names = "--save-json")
    private String jsonFile;
    @Parameter(names = "--color")
    private boolean color;

    static class Point {
        double lat, lng;
        String name;
        Point(double lat, double lng, String name) { this.lat=lat; this.lng=lng; this.name=name; }
    }

    static class Route {
        Point start, end;
        List<Point> waypoints;
        double totalDistanceKm;
        int elevationGainM;
        double timeHours;
        String profile;
        List<Point> points;
    }

    private Point parseLocation(String loc) {
        if (loc.contains(",")) {
            String[] parts = loc.split(",");
            try {
                double lat = Double.parseDouble(parts[0].trim());
                double lng = Double.parseDouble(parts[1].trim());
                return new Point(lat, lng, loc);
            } catch (NumberFormatException ignored) {}
        }
        return new Point(0, 0, loc);
    }

    private Point geocode(String name) {
        int hash = name.hashCode() % 360;
        if (hash < 0) hash += 360;
        return new Point(50 + (hash % 100) / 100.0, 10 + (hash % 200) / 100.0, name);
    }

    private double haversine(double lat1, double lon1, double lat2, double lon2) {
        final double R = 6371.0;
        double phi1 = Math.toRadians(lat1);
        double phi2 = Math.toRadians(lat2);
        double dphi = Math.toRadians(lat2 - lat1);
        double dlambda = Math.toRadians(lon2 - lon1);
        double a = Math.sin(dphi/2)*Math.sin(dphi/2) + Math.cos(phi1)*Math.cos(phi2)*Math.sin(dlambda/2)*Math.sin(dlambda/2);
        return R * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1-a));
    }

    private double getSpeed() {
        switch (profile) {
            case "road": return 25;
            case "mtb": return 15;
            default: return 20;
        }
    }

    private Route calculateRoute() {
        Point startP = parseLocation(start);
        Point endP = parseLocation(end);
        List<Point> wayPts = new ArrayList<>();
        if (waypoints != null) {
            for (String w : waypoints.split(",")) wayPts.add(parseLocation(w.trim()));
        }
        if (startP.lat == 0) startP = geocode(startP.name);
        if (endP.lat == 0) endP = geocode(endP.name);
        for (Point w : wayPts) {
            if (w.lat == 0) { Point g = geocode(w.name); w.lat=g.lat; w.lng=g.lng; }
        }
        List<Point> points = new ArrayList<>();
        points.add(startP);
        points.addAll(wayPts);
        points.add(endP);
        double totalDist = 0;
        for (int i = 0; i < points.size()-1; i++) {
            totalDist += haversine(points.get(i).lat, points.get(i).lng, points.get(i+1).lat, points.get(i+1).lng);
        }
        double speed = getSpeed();
        double timeHours = totalDist / speed;
        int elevGain = (int)(totalDist * 0.05 * 1000);
        Route route = new Route();
        route.start = startP;
        route.end = endP;
        route.waypoints = wayPts;
        route.totalDistanceKm = totalDist;
        route.elevationGainM = elevGain;
        route.timeHours = timeHours;
        route.profile = profile;
        route.points = points;
        return route;
    }

    private void printRoute(Route r) {
        if (color || System.console() != null) {
            System.out.println("\u001B[36m🚴 Маршрут от " + r.start.name + " до " + r.end.name + "\u001B[0m");
            if (!r.waypoints.isEmpty()) {
                String names = String.join(", ", r.waypoints.stream().map(w -> w.name).toArray(String[]::new));
                System.out.println("\u001B[35m   через " + names + "\u001B[0m");
            }
            System.out.printf("\u001B[32mРасстояние: %.1f км\u001B[0m\n", r.totalDistanceKm);
            System.out.printf("\u001B[33mНабор высоты: %d м\u001B[0m\n", r.elevationGainM);
            System.out.printf("\u001B[34mВремя в пути: ~%.1f ч\u001B[0m\n", r.timeHours);
        } else {
            System.out.println("Маршрут от " + r.start.name + " до " + r.end.name);
            if (!r.waypoints.isEmpty()) {
                String names = String.join(", ", r.waypoints.stream().map(w -> w.name).toArray(String[]::new));
                System.out.println("   через " + names);
            }
            System.out.printf("Расстояние: %.1f км\n", r.totalDistanceKm);
            System.out.printf("Набор высоты: %d м\n", r.elevationGainM);
            System.out.printf("Время в пути: ~%.1f ч\n", r.timeHours);
        }
    }

    private void exportGPX(Route r, String filename) throws IOException {
        StringBuilder gpx = new StringBuilder();
        gpx.append("<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n");
        gpx.append("<gpx version=\"1.1\" creator=\"BikeRoutePlanner\">\n");
        gpx.append("  <trk>\n    <trkseg>\n");
        for (Point pt : r.points) {
            gpx.append(String.format("      <trkpt lat=\"%.6f\" lon=\"%.6f\">\n", pt.lat, pt.lng));
            gpx.append(String.format("        <ele>%d</ele>\n", (int)(50 + pt.lat * 10)));
            gpx.append("      </trkpt>\n");
        }
        gpx.append("    </trkseg>\n  </trk>\n</gpx>");
        Files.write(Paths.get(filename), gpx.toString().getBytes());
        System.out.println("GPX экспортирован в " + filename);
    }

    private void saveJSON(Route r, String filename) throws IOException {
        Gson gson = new GsonBuilder().setPrettyPrinting().create();
        Files.write(Paths.get(filename), gson.toJson(r).getBytes());
        System.out.println("Маршрут сохранён в " + filename);
    }

    public static void main(String[] args) throws Exception {
        BikeRoutePlanner planner = new BikeRoutePlanner();
        JCommander.newBuilder().addObject(planner).build().parse(args);
        Route route = planner.calculateRoute();
        planner.printRoute(route);
        if (planner.gpxFile != null) planner.exportGPX(route, planner.gpxFile);
        if (planner.jsonFile != null) planner.saveJSON(route, planner.jsonFile);
    }
}
